[slimtensor] Storage and SharedPtr for CPU owning mode

Gasoonjia opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #16385
* #16384
* __->__ #16383
* #16382

This diff introduces the foundation for tensor storage management in the SlimTensor migration:

- util/SharedPtr.h - A lightweight, non-thread-safe shared pointer optimized for single-threaded tensor operations.
- core/Storage.h - The MaybeOwningStorage class that manages tensor data memory:
    - DeviceTraits<CPU> specialization with allocate(), free(), memcpy() using malloc/free
    - Owning mode for CPU device (CUDA and non-owning mode added later)
    - Storage type alias as SharedPtr<MaybeOwningStorage>
    - Move semantics for efficient resource transfer
    - clone() and copy_() methods for data management

Differential Revision: [D89747980](https://our.internmc.facebook.com/intern/diff/D89747980/)

Gasoonjia committed:
[slimtensor] Storage and SharedPtr for CPU owning mode

This diff introduces the foundation for tensor storage management in the SlimTensor migration:

- util/SharedPtr.h - A lightweight, non-thread-safe shared pointer optimized for single-threaded tensor operations.
- core/Storage.h - The MaybeOwningStorage class that manages tensor data memory:
    - DeviceTraits<CPU> specialization with allocate(), free(), memcpy() using malloc/free
    - Owning mode for CPU device (CUDA and non-owning mode added later)
    - Storage type alias as SharedPtr<MaybeOwningStorage>
    - Move semantics for efficient resource transfer
    - clone() and copy_() methods for data management

Differential Revision: [D89747980](https://our.internmc.facebook.com/intern/diff/D89747980/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[slimtensor] Storage and SharedPtr for CPU owning mode"


This diff introduces the foundation for tensor storage management in the SlimTensor migration:

- util/SharedPtr.h - A lightweight, non-thread-safe shared pointer optimized for single-threaded tensor operations.
- core/Storage.h - The MaybeOwningStorage class that manages tensor data memory:
    - DeviceTraits<CPU> specialization with allocate(), free(), memcpy() using malloc/free
    - Owning mode for CPU device (CUDA and non-owning mode added later)
    - Storage type alias as SharedPtr<MaybeOwningStorage>
    - Move semantics for efficient resource transfer
    - clone() and copy_() methods for data management

Differential Revision: [D89747980](https://our.internmc.facebook.com/intern/diff/D89747980/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[slimtensor] Storage and SharedPtr for CPU owning mode"


This diff introduces the foundation for tensor storage management in the SlimTensor migration:

- util/SharedPtr.h - A lightweight, non-thread-safe shared pointer optimized for single-threaded tensor operations.
- core/Storage.h - The MaybeOwningStorage class that manages tensor data memory:
    - DeviceTraits<CPU> specialization with allocate(), free(), memcpy() using malloc/free
    - Owning mode for CPU device (CUDA and non-owning mode added later)
    - Storage type alias as SharedPtr<MaybeOwningStorage>
    - Move semantics for efficient resource transfer
    - clone() and copy_() methods for data management

Differential Revision: [D89747980](https://our.internmc.facebook.com/intern/diff/D89747980/)

[ghstack-poisoned]

JacobSzwejbka reviewed (changes requested):
Review automatically exported from Phabricator review in Meta.

Gasoonjia committed:
Update on "[slimtensor] Storage and SharedPtr for CPU owning mode"


This diff introduces the foundation for tensor storage management in the SlimTensor migration:

- util/SharedPtr.h - A lightweight, non-thread-safe shared pointer optimized for single-threaded tensor operations.
- core/Storage.h - The MaybeOwningStorage class that manages tensor data memory:
    - DeviceTraits<CPU> specialization with allocate(), free(), memcpy() using malloc/free
    - Owning mode for CPU device (CUDA and non-owning mode added later)
    - Storage type alias as SharedPtr<MaybeOwningStorage>
    - Move semantics for efficient resource transfer
    - clone() and copy_() methods for data management

Differential Revision: [D89747980](https://our.internmc.facebook.com/intern/diff/D89747980/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[slimtensor] Storage and SharedPtr for CPU owning mode"


This diff introduces the foundation for tensor storage management in the SlimTensor migration:

- util/SharedPtr.h - A lightweight, non-thread-safe shared pointer optimized for single-threaded tensor operations.
- core/Storage.h - The MaybeOwningStorage class that manages tensor data memory:
    - DeviceTraits<CPU> specialization with allocate(), free(), memcpy() using malloc/free
    - Owning mode for CPU device (CUDA and non-owning mode added later)
    - Storage type alias as SharedPtr<MaybeOwningStorage>
    - Move semantics for efficient resource transfer
    - clone() and copy_() methods for data management

Differential Revision: [D89747980](https://our.internmc.facebook.com/intern/diff/D89747980/)

[ghstack-poisoned]

Gasoonjia committed:
Update on "[slimtensor] Storage and SharedPtr for CPU owning mode"


This diff introduces the foundation for tensor storage management in the SlimTensor migration:

- util/SharedPtr.h - A lightweight, non-thread-safe shared pointer optimized for single-threaded tensor operations.
- core/Storage.h - The MaybeOwningStorage class that manages tensor data memory:
    - DeviceTraits<CPU> specialization with allocate(), free(), memcpy() using malloc/free
    - Owning mode for CPU device (CUDA and non-owning mode added later)
    - Storage type alias as SharedPtr<MaybeOwningStorage>
    - Move semantics for efficient resource transfer
    - clone() and copy_() methods for data management

Differential Revision: [D89747980](https://our.internmc.facebook.com/intern/diff/D89747980/)

[ghstack-poisoned]
